pkg/reconciler/timeout_handler_test.go

```diff
@@ -259,3 +259,80 @@ func TestWithNoFunc(t *testing.T) {
 	testHandler.CheckTimeouts()
 
 }
+
+// TestGetTaskRunEndTime checks that the end time calculated for a given taskrun is consistent
+// and correctly uses the default taskrun timeout.
+func TestGetTaskRunEndTime(t *testing.T) {
+	timeNow := time.Now()
+	timeInTwoSeconds := timeNow.Add(2 * time.Second)
+	timeAfterDefaultTimeout := timeNow.Add(defaultTimeout)
+	testCases := []struct {
+		startTime       time.Time
+		optionalTimeout time.Duration
+		expectedEndTime time.Time
+	}{
+		{
+			startTime:       timeNow,
+			optionalTimeout: 2 * time.Second,
+			expectedEndTime: timeInTwoSeconds,
+		},
+		{
+			startTime:       timeNow,
+			expectedEndTime: timeAfterDefaultTimeout,
+		},
+	}
+	for i, tc := range testCases {
+		taskRun := tb.TaskRun("test-taskrun-get-endtime", testNs, tb.TaskRunSpec(
+			tb.TaskRunTaskRef(simpleTask.Name, tb.TaskRefAPIVersion("a1")),
+		), tb.TaskRunStatus(tb.Condition(apis.Condition{
+			Type:   apis.ConditionSucceeded,
+			Status: corev1.ConditionUnknown}),
+			tb.TaskRunStartTime(tc.startTime),
+		))
+		if tc.optionalTimeout != time.Duration(0) {
+			tb.TaskRunTimeout(tc.optionalTimeout)(&taskRun.Spec)
+		}
+		endTime := GetTaskRunEndTime(taskRun)
+		if !endTime.Equal(tc.expectedEndTime) {
+			t.Fatalf("testcase %d expected %q received %q", i+1, tc.expectedEndTime.Format(time.StampNano), endTime.Format(time.StampNano))
```

Comment:
nit: if you make these subtests (use t.Run and t.Fail) instead of t.Fatal, then the test runner will run them all even if the first one fails.